lib/Sema/CSDiagnostics.h

```diff
@@ -81,6 +81,8 @@ class FailureDiagnostic {
 
   DeclContext *getDC() const { return getConstraintSystem().DC; }
 
+  const Solution &getSolution() const { return solution; }
```

Comment:
I'd prefer that nothing could access to solution directly but instead had accessors for some of the stuff in it e.g. the way it's used in the code could be extracted into `findRestriction`...